pertpy/tools/_distances.py

```diff
@@ -0,0 +1,407 @@
+    if metric == "edistance":
+        return Edistance()(x, y)
+    elif metric == "wasserstein":
+        return Wasserstein()(x, y)
+
+
+def get_distances(
+    adata: AnnData,
+    groups: list[str],
+    metric: str | Metric = "edistance",
+    **kwargs,
+) -> pd.DataFrame:
+    """Get pairwise distances between groups of cells.
+
+    Args:
+        adata (AnnData): Annotated data matrix.
+        groups (List[str]): List of group names. # or None, then use all groups
+        metric (Union[str, Metric], optional): Distance metric. Defaults to 'edistance'.
+
+    Returns:
+        pd.DataFrame: Dataframe with pairwise distances.
+    """
+    raise NotImplementedError("get_distances is not implemented yet.")
+
+
+class Metric:
+    """Metric class of distance metrics between groups of cells. (OUTDATED)"""
+
+    def __call__(self, X: np.ndarray, Y: np.ndarray, **kwargs) -> float:
+        """Compute distance between vectors X and Y.
+
+        Args:
+            X (np.ndarray): First vector of shape (n_samples, n_features).
+            Y (np.ndarray): Second vector of shape (n_samples, n_features).
+
+        Returns:
+            float: Distance between X and Y.
+        """
+        raise NotImplementedError("Metric class is abstract.")
+
+    def from_Pidx(self, P: np.ndarray, idx: np.ndarray, **kwargs) -> float:
+        """Compute a distance between vectors X and Y with precomputed distances.
+
+        Args:
+            P (np.ndarray): Pairwise distance matrix of shape (n_samples, n_samples).
+            idx (np.ndarray): Boolean array of shape (n_samples,) indicating which
+            samples belong to X (or Y, since each metric is symmetric).
+
+        Returns:
+            float: Distance between X and Y.
+        """
+        raise NotImplementedError("Metric class is abstract.")
+
+
+class Edistance(Metric):
+    """Edistance metric."""
+
+    def __call__(self, X: np.ndarray, Y: np.ndarray, **kwargs) -> float:
+        # TODO: inherit docstring from parent class
+        sigma_X = pairwise_distances(X, X, metric="euclidean").mean()
+        sigma_Y = pairwise_distances(Y, Y, metric="euclidean").mean()
+        delta = pairwise_distances(X, Y, metric="euclidean").mean()
+        return 2 * delta - sigma_X - sigma_Y
+
+    def from_Pidx(self, P: np.ndarray, idx: np.ndarray, **kwargs) -> float:
+        # TODO: inherit docstring from parent class
+        sigma_X = P[idx, :][:, idx].mean()
+        sigma_Y = P[~idx, :][:, ~idx].mean()
+        delta = P[idx, :][:, ~idx].mean()
+        return 2 * delta - sigma_X - sigma_Y
+
+
+class MMD(Metric):
+    # TODO: implement MMD metric
+    # e.g.
+    # https://github.com/jindongwang/transferlearning/blob/master/code/distance/mmd_numpy_sklearn.py
+    # https://github.com/calico/scmmd
+    pass
+
+
+class Wasserstein(Metric):
+    """Wasserstein distance metric.
+    NOTE: I bet the Theislab OT people have a better solution for this.
```

Comment:
There should be a Jax implementation here https://github.com/theislab/moscot

Reply:
WIP: https://github.com/theislab/moscot/issues/493

Reply:
I have some issues with moscot atm. Might take ma a while to figure out how to use it.